build(docs-infra): simplify configuring redirects

gkalpak opened the pull request:
~~NOTE: This PR sits on top of #42414 to avoid rebase conflicts.~~ _#42414 has been merged._

##
This PR cleans up the configured Firebase and ServiceWorker redirects and the corresponding tests and simplifies the process of configuring redirects by auto-generating the ServiceWorker config `navigationUrls` from the Firebase redirects config.

See individual commits for more details.

mary-poppins commented:
You can preview 689feae at https://pr42452-689feae.ngbuilds.io/.

IgorMinar reviewed (approved):
This looks great! thanks @gkalpak

gkalpak committed:
test(docs-infra): add testcases for all redirect rules

This commit adds testcases in `URLS_TO_REDIRECT.txt` for all redirect
rules defined in `firebase.json`. This ensures that all rules are tested
and work as expected.

gkalpak committed:
build(docs-infra): add missing redirect rule for `api/http` in `firebase.json`

Previously, only sub-paths, such as `api/http/foo`, were redirected to
`guide/deprecations#http`.

This commit ensures that also `api/http` itself (which used to point to
the `http` module's API page) is redirected as well.

gkalpak committed:
test(docs-infra): ensure all redirect rules are tested

This commit adds a test assertion to verify that all redirect rules
defined in `firebase.json` are tested, i.e. that each rule is applied to
at least one testcase from `URLS_TO_REDIRECT.txt`.

This will ensure that any redirect rules added in the future will be
tested.

gkalpak committed:
test(docs-infra): unnest unnecessarily nested test suite

Previously, the `destinations` test suite in
`testFirebaseRedirection.spec.ts` was nested under the
`with legacy URLs` test suite. However, the two suites are unrelated and
there is no reason to have them nested.

This commit moves the `destinations` test suite to be beside (instead of
inside) the `with legacy URLs` one.

gkalpak committed:
build(docs-infra): remove redundant `navigationUrls` globs from `ngsw-config.json`

This commit removes some `navigationUrls` globs from `ngsw-config.json`
that have no effect. More specifically:
- It removes globs of the form `!/**/xyz.html`, since these are already
  covered by the more generic `!/**/*.*` glob.
- It removes `!/api/**/NgFor`, since it does not have a corresponding
  redirect rule in `firebase.json`.

This is in preparation of automatically generating the ServiceWorker
`navigationUrls` based on `firebase.json`.

gkalpak committed:
build(docs-infra): allow a trailing `/` in more `navigationUrls` globs in `ngsw-config.json`

Previously, we were not consistent in allowing a trailing `/` in
`navigationUrls` globs in `ngsw-config.json`. And when we did, we used
two globs: one with and one without the trailing `/`.

This commit updates all appropriate `navigationUrls` globs to allow a
trailing `/`. It also merges the two glob patterns (the one with and the
one without the `/`) into one.

This is in preparation of automatically generating the ServiceWorker
`navigationUrls` based on `firebase.json`.

gkalpak committed:
build(docs-infra): alphabetically sort all `navigationUrls` in `ngsw-config.json`

This commit sorts all `navigationUrls` in `ngsw-config.json`
alphabetically.

This is in preparation of automatically generating the ServiceWorker
`navigationUrls` based on `firebase.json`.

gkalpak committed:
build(docs-infra): align `navigationUrls` in `ngsw-config.json` with Firebase redirects

The ServiceWorker `navigationUrls` globs defined in `ngsw-config.json`
are supposed to exclude any URLs that are redirected on the server (as
configured in `firebase.json`). However, the list of redirected
URLs/globs in `firebase.json` and `ngsw-config.json` have gotten out of
sync.

This commit updates the globs in `ngsw-config.json` to match the ones in
`firebase.json`.

This is in preparation of automatically generating the ServiceWorker
`navigationUrls` based on `firebase.json`.

gkalpak committed:
build(docs-infra): auto-generate SW `navigationUrls` from Firebase config

Previously, redirects had to be configured in both the Firebase config
(`firebase.json`) and the ServiceWorker config (`ngsw-config.json`).
This made it challenging to correctly configure redirects, since one had
to understand the different formats of the two configs, and was also
prone to getting out-of-sync configs.

This commit simplifies the process of adding redirects by removing the
need to update the ServiceWorker config (`ngsw-config.json`) and keep it
in sync with the Firebase config (`firebase.json`). Instead the
ServiceWorker `navigationUrls` are automatically generated from the list
of redirects in the Firebase config.

NOTE:
Currently, the automatic generation only supports the limited set of
patterns that are necessary to translate the existing redirects. It can
be made more sophisticated in the future, should the need arise.

gkalpak committed:
docs(docs-infra): document how to configure redirects

mary-poppins commented:
You can preview 00e6217 at https://pr42452-00e6217.ngbuilds.io/.
